src/v/resource_mgmt/storage.cc

```diff
@@ -100,86 +101,306 @@ ss::future<> disk_space_manager::run_loop() {
-        // make sure housekeeping doesn't delete the log from below us
-        auto gate = log->gate().hold();
+    gc_config cfg(
+      collection_threshold(),
+      _storage->local().log_mgr().config().retention_bytes());
 
-        auto segments = log->cloud_gc_eligible_segments();
+    /*
+     * in smallish batches partitions are queried for their reclaimable
+     * segments. all of this information is bundled up and returned.
+     */
+    fragmented_vector<partition> res;
+    co_await ss::max_concurrent_for_each(
+      partitions.begin(), partitions.end(), 20, [&res, cfg](const auto& p) {
+          auto log = dynamic_cast<storage::disk_log_impl*>(p->log().get_impl());
+          auto gate = log->gate().hold(); // protect against log deletes
+          return log->get_reclaimable_offsets(cfg)
+            .then([&res, p](auto offsets) {
+                res.push_back({
+                  .group = p->group(),
+                  .offsets = std::move(offsets),
+                });
+            })
+            .finally([g = std::move(gate)] {});
+      });
 
-        vlog(
-          rlog.info,
-          "Remote partition {} reports {} reclaimable segments",
-          p->ntp(),
-          segments.size());
+    /*
+     * sorting by raft::group_id would provide a more stable round-robin
+     * evaluation of partitions in later processing of the result, but on small
+     * timescales we don't expect enough churn to make a difference, nor with a
+     * large number of partitions on the system would the non-determinism
+     * in the parallelism above in max_concurrent_for_each be problematic.
+     */
 
-        if (segments.empty()) {
-            continue;
-        }
+    vlog(rlog.trace, "Reporting reclaim offsets for {} partitions", res.size());
+    co_return res;
+}
+
+ss::future<> eviction_policy::install_schedule(schedule schedule) {
+    /*
+     * install the schedule on each core in parallel
+     */
+    auto total = co_await ss::map_reduce(
+      schedule.shards,
+      [this](auto& shard) { return install_schedule(std::move(shard)); },
+      size_t(0),
+      std::plus<>());
+
+    vlog(rlog.info, "Requested truncation from {} cloud partitions", total);
+}
+
+// the per-shard counterpart of install_schedule(schedule)
+ss::future<size_t> eviction_policy::install_schedule(shard_partitions shard) {
+    const auto shard_id = shard.shard;
+    return _pm->invoke_on(shard_id, [shard = std::move(shard)](auto& pm) {
+        size_t decisions = 0;
+        for (const auto& partition : shard.partitions) {
+            if (!partition.decision.has_value()) {
+                continue;
+            }
+
+            auto p = pm.partition_for(partition.group);
+            if (!p) {
+                continue;
+            }
+
+            auto log = dynamic_cast<storage::disk_log_impl*>(
+              p->log().get_impl());
+            log->set_cloud_gc_offset(partition.decision.value());
```

Comment:
At a high-level the housekeeping loop _actually_ removes data, and the space manager only makes decisions. It's separated out because the space manager will likely also be controlling aspects of the cloud cache too at some point.

The general structure of housekeeping is:

1. do some GC do some compaction (same as before)
2. if low on disk OR an explicit GC offset has been set, prioritize GC over compaction

the interaction then with space management is that space management will set target GC offsets on partitions during scheduling, and then the last step is to trigger housekeeping which will then prioritize a GC pass.

I think it is a little unfortunate that so much of the storage internals pops up at the level of the space management control loop. One of the original PRs had pushed this down into log_manager, but it dragged along a bunch of weird dependencies that were circular. I think there is more to come on combining or refactoring.

Reply:
>if low on disk OR an explicit GC offset has been set

Thanks! I'd forgotten this was the case.